packages/publisher/index.ts

```diff
@@ -0,0 +1,226 @@
+                    .toString()
+    }
+    return null;
+}
+
+const BASE_URL = new URL('https://discord.com/api/v10/applications/');
+const PUBLISHABLE = 0b1110;
+
+export class Publisher implements Init {
+    constructor(private modules: Map<string, CommandModule>,
+                private sernEmitter : Emitter,
+                private logger: Logging) {}
+
+    async init() {
+        if(!process.env.DISCORD_TOKEN) {
+            throw Error("No token found to publish. add DISCORD_TOKEN to .env");
+        }
+        const headers = [['Authorization', 'Bot ' + process.env.DISCORD_TOKEN],
+                         ['Content-Type', 'application/json']] as Array<[string,string]>
+        let me; 
+        let appid: string;
+        try {
+            me = await fetch(new URL('@me', BASE_URL), { headers }).then(res => res.json());
+            appid = me.id;
+        } catch(e) {
+            console.log("Something went wrong while trying to fetch your application:");
+            throw e;
+        }
+        const GLOBAL_URL = new URL(`${appid}/commands`, BASE_URL);
+        interface LocalPublish {
+            guildIds?: string[]
+            default_member_permissions: string,
+            integration_types: string[],
+            contexts: number[]
+        }
+        const listener = async () => {
+            this.logger.info({ message: 'publishing modules' });
+            const modules = 
+                Array.from(this.modules.values())
+                     .filter(module => (module.type & PUBLISHABLE) != 0)
+                     .map(module => {
+                        const publish = module.locals.publish as LocalPublish || {}
+                        return {
+                            guildIds: publish?.guildIds ?? [],
+                            toJSON() {
+                                const applicationType = intoApplicationType(module.type);
+                                const { default_member_permissions,    
+                                        integration_types,
+                                        contexts } = publish;
+                                return {
+                                    name: module.name, type: applicationType,
+                                    //@ts-ignore we know description is at least empty str or filled
+                                    description: makeDescription(applicationType, module.description),
+                                    //@ts-ignore shutup
+                                    options: optionsTransformer(module?.options),
+                                    default_member_permissions,
+                                    integration_types, contexts,
+                                    //@ts-ignore
+                                    name_localizations: module.locals.nloc, 
+                                    //@ts-ignore
+                                    description_localizations: module.locals.dloc
```

Comment:
thats a lot of domination over typescript i see 💀

Reply:
i should probably add more typings to make it stop complaining